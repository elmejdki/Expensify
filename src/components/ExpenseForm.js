import React, { Component } from 'react';
import moment from 'moment';
import 'react-dates/initialize';
import { SingleDatePicker } from 'react-dates';

export default class ExpenseForm extends Component {
  state = {
    description: this.props.description || '',
    note: this.props.note || '',
    amount: this.props.amount ? this.props.amount / 100 + '' : '',
    createdAt: this.props.createdAt ? moment(this.props.createdAt) : moment(),
    calendarFocused: false,
    state: ''
  }

  onDescriptionChange = (event) => {
    this.setState(() => ({
      description: event.target.value
    }));
  }

  onNoteChange = (event) => {
    this.setState(() => ({
      note: event.target.value
    }))
  }

  onAmountChange = (event) => {
    const amount = event.target.value;
    if(!amount || amount.match(/^\d+(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  }

  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  }

  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  }

  onSubmit = (e) => {
    e.preventDefault();
    if (!this.state.description || !this.state.amount) {
      // set error state equal to 'Please provide description and amount'
      this.setState(() => ({
        error: 'Please provide description and amount'
      }));
    }else {
      this.setState(() => ({
        error: ''
      }));
      const { description, note, amount, createdAt } = this.state;
      this.props.onSubmit({
        description,
        note,
        createdAt: createdAt.valueOf(),
        amount: parseInt(Number(amount)*100)  
      });
    }
  }

  render() {
    return (
      <form className="form" onSubmit={this.onSubmit}>
        { this.state.error && <p className="form__error">{this.state.error}</p>}
        <input 
          type="text"
          className="text-input"
          placeholder="Description"
          autoFocus
          value={this.state.description}
          onChange={this.onDescriptionChange}
        />
        <input
          type="text"
          className="text-input"
          placeholder="amount"
          value={this.state.amount}
          onChange={this.onAmountChange}
        />
        <SingleDatePicker
          date={this.state.createdAt}
          onDateChange={this.onDateChange}
          focused={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
        <textarea
          value={this.state.note}
          className="textarea"
          onChange={this.onNoteChange}
          placeholder="Add a note for your expense (optional)"
        >
        </textarea>
        <div>
          <button className="button">
            {
              this.props.isEdit ? 'Edit Expense' : 'Add Expense'
            }
          </button>
        </div>
      </form>
    );
  }
}
