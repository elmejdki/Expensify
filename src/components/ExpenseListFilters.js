import React, { Component } from 'react';
import { DateRangePicker } from 'react-dates';
import { connect } from 'react-redux';
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters';

class ExpenseListFilters extends Component {
  state = {
    calendarFocused: null
  };

  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  }

  onFocusChange = (calendarFocused) => {
    this.setState(() => ({ calendarFocused }));
  }

  getDateId = (date) => {
    return date ? date.format('Do MMMM, YYYY') : 'empty';
  }

  render() {
    const { filters, setTextFilter, sortByAmount, sortByDate } = this.props;
    const { text, sortBy, endDate, startDate } = filters;

    return (
      <div className="content-container">
        <div className="input-group">
          <div className="input-group__item">
            <input
              className="text-input"
              placeholder="Search expenses"
              type="text"
              value={text}
              onChange={(e) => {
                setTextFilter(e.target.value);
              }}
            />
          </div>
          <div className="input-group__item">
            <select
              value={sortBy}
              className="select"
              onChange={(e) => {
                if(e.target.value === 'amount') {
                  sortByAmount();
                }else if(e.target.value === 'date') {
                  sortByDate();
                }
              }}
            >
              <option value="date">Date</option>
              <option value="amount">Amount</option>
            </select>
          </div>
          <div className="input-group__item">
            <DateRangePicker
              startDate={startDate}
              startDateId={this.getDateId(startDate)}
              endDate={endDate}
              endDateId={this.getDateId(endDate)}
              onDatesChange={this.onDatesChange}
              focusedInput={this.state.calendarFocused}
              onFocusChange={this.onFocusChange}
              showClearDates={true}
              numberOfMonths={1}
              isOutsideRange={() => false}
            />
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    filters: state.filters
  };
}

const mapDispatchToProps = {
  setTextFilter,
  sortByAmount,
  sortByDate,
  setStartDate,
  setEndDate
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ExpenseListFilters);