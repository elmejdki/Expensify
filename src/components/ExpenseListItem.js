import React from 'react';
import { connect } from 'react-redux';
import { removeExpense } from '../actions/expenses';

const ExpenseListItem = ({ id, description, amount, createdAt, removeExpense }) => (
  <div>
    <h3>{description}</h3>
    <p>
      {amount} - {
        new Date(createdAt)
          .toString()
          .match(/^\w{3}(\s*.)+\s\d{4}/)[0] }
    </p>
    <button
      onClick={() => {
        removeExpense(id); 
      }}
    >
      Remove
    </button>
  </div>
);

const mapDispatchToProps = { removeExpense };

export default connect(null, mapDispatchToProps)(ExpenseListItem);