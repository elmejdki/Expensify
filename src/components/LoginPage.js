import React from 'react';
import { connect } from 'react-redux';
import { startLogin } from '../actions/auth';

export const LoginPage = ({ startLogin }) => (
  <div>
    <button onClick={startLogin}>Log In</button>
  </div>
);

const mapDispatchToProps = { startLogin };

export default connect(undefined, mapDispatchToProps)(LoginPage);
