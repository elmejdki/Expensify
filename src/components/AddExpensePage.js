import React from 'react';
import { connect } from 'react-redux';
import { addExpense } from '../actions/expenses';
import ExpenseForm from './ExpenseForm';

const AddExpensePage = (props) => (
  <div>
    <h1>Add Expense</h1>
    <ExpenseForm
      onSubmit={(expense) => {
        props.addExpense(expense);
        props.history.push('/');
      }}
    />
  </div>
);

const mapDispatchToProps = { addExpense };

export default connect(null, mapDispatchToProps)(AddExpensePage);