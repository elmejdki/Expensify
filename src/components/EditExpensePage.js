import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';

export const EditExpensePage = (props) => {
  const id = props.match.params.id;

  return (
    <div>
      <ExpenseForm
        {...props.expense}
        onSubmit={(expense) => {
          props.startEditExpense(id, expense);
          props.history.push('/dashboard');
        }}
        isEdit={true}
      />
      <button
        onClick={() => {
          props.startRemoveExpense(id);
          props.history.push('/dashboard');
        }}
      >
        Remove
      </button>
    </div>
  )
};

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find(expense => expense.id === props.match.params.id)
  };
}

const mapDispatchToProps = { startEditExpense, startRemoveExpense };

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(EditExpensePage);