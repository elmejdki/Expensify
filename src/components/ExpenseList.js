import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

const ExpenseList = (props) => (
  <div>
    <h1>Expense List</h1>
    {props.expenses.map(expense => (
      <ExpenseListItem key={expense.id} {...expense} />
    ))}
  </div>
);

const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters)
  };
};

const ConnectedExpenseList = connect(mapStateToProps)(ExpenseList);

export default ConnectedExpenseList;
