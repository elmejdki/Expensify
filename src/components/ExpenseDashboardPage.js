import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';

const ExpenseDashboardPage = () => (
  <div>
    <ExpenseListFilters />
    <ExpenseList />
  </div>
);

export default ExpenseDashboardPage;