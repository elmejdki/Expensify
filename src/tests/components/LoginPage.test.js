import ShallowRenderer from 'react-test-renderer/shallow';
import React from 'react';
import { LoginPage } from '../../components/LoginPage';

const renderer = new ShallowRenderer();

it('should render LoginPage correctly', () => {
  renderer.render(<LoginPage />);

  expect(renderer.getRenderOutput()).toMatchSnapshot();
});
