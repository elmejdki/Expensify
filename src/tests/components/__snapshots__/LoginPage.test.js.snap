// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render LoginPage correctly 1`] = `
<div>
  <button>
    Log In
  </button>
</div>
`;
